### ISSUE TITLE
Strange rendering of gocesiumtiler tiles with custom Cesium JS shader

### ISSUE THREAD

**Issue by daz** [mfbonfigli/gocesiumtiler#47]
I am rendering a point cloud in Cesium JS with a custom shader, it’s displaying perfectly in the [sandcastle](https://sandcastle.cesium.com/#c=lVZZb9s4EP4rRLAP8q5CS5Z85GixgesmQZI2iN30xYBDS7TNLUV6KcpHivz3HZKS4yMtsjZgkxzO/c1w6vV7yYRGXS6LFI3X6F7JCUb3VFOFLjinwkddyYtszAj6JtiCqpzpNXqQY6lZkqNbMsaonxAhmJgaARec/kNEqiQaKJkkkjMEW3RHNGWoy2RCFGcUD0UiRa7RgtElqPqABF2iLs1ZkeFHe+YNjxK770qhCRNUDY989HMoEALjFJycVgwDt8cTJbPvUvG0PPBq/lC81M6GYigWRKEZWML3tfUTRanoz0lCewsq9JW75DnLcJ5QQTE4uCC5kVTKwDnV12Je6ItEMym8SSHcghoZNWcnQkYreKxBFRGgd0fonCU/7iXEc8OH5+XWaDL8JkiSU8zl1NvIMcQX/1f2D9Zzim97nwej7u1198Z5D/7vupNRRYwPJtaeNTaluWaCGO2n2/HpVmojz9kUx0HYjmPciuMwbkZh7LvzRqMZNEMch0G7cxKGQbNREuLmSTNuBDhqx+1OHEeROa5ZolQMTD5UekVJCoC6ZzqZPUjOS9VNUNpqR1EnaAbRSbvZLjUcB7jRbrWCdtiMOtFJp9MqCS3c6IRgS6fdjIMg6DQr1S9VXMwXlR+TrYki0wxs6s9IStWArjSk7QniJ1m6Id4ZcH0uNxYGaJLbfx8xIWGbPGejTKaUG+ArRjjKykVtKCDeQ13tccomkyKnoCZjK28xSlk+52Rt2Ue5XnMIxAgKSSof7TP5KMChxcNQPJ3t+wLVqulq3xNXeBMuiUaXDxefrntfBqOr3vXl1QAuRAEOKsjQJEKzfNFQ07FnN4kF9lDDJkY3Bs+w8EIc+KiBA1RHkVmGr0v4sYHWln0OHGScexDGRHsJXq1W6C90g1fr5xr6E5IVoGPYLpdLx6SoLpRACX4GqgnOjWHxUcJJNvfm9rI9CEq1NaDhdRmPMmcuCDZjj3bp8rX4Rb7cna+Fdpfcoqxn64SiE6qoSKjtnHddF4XIszbYn2ZQel2xVFXtLjvFmGit2LiA2sKv9JLLJQeAoAkoKqO2rxiicrwlGj/XdrhnhWH0wCmvEuTv57tWg0TtnZ29ejodg4RtAJhUg1IQbcPtYu561e9wC1JAwllVbFVUwW9wpM+ejaENC7sDGDuwJkWuZeaAvNu9u1sU7+crH0KcTWfQ0KZ3Jqubp+J2+xR/+3J7PfC3uQrBJlJl+SnaEQaE0XKad+Kvgn4FRD1Ae2L9fwuiaHpwFR4o6MEbld+cSNuXH3vdyN+/vSC8eL3e45zNc0Au/n7Z78RYHmrck/Cys9/dQRtYg7uH/tTrqO9yhFyO9AxAw3J4rnhScOgzKZRGWT0od6E3LzmZz+EBt8SqHZbkXQW/A8TG2Udn3tux2fNkr5mdHpzsXD/s46dvnG2xlC+CVmsXKoc8zTiFZ9KU4JIwvYGd/Ys+DRzZTh7XUlzkZixIvbAVN0JXF6UAvAfh7e2ZScb58f/8lInIEUFm1hnLJZoqAAn4Z/TW65VmE3m6VzVb5hsUUM9eqr3bki3la8o56LblhpZMzwAYgFFb6DvTjmIZtKoF9DuSpl5pne3VgDl455EHU5tUtdfwV5PPU88QELQ1MxRUIT1Ff/y0HC9PTsq7xpx3DTqHI85b083hYFMr8fTuseZwoHlrljkcY2rlBHPkH53b1H2sgPw3y+ZSaVQo7mFc1zSbm2LO6+Mi+WFwmOfVcHle32Y9T9kCsfTDG2O3eXDzHCiTgnPTr4dHH8/rcP+AtUyRaVhQ+ebaLPx46w4xxud12L7NqaXkY6L2JP8H) with the demo point cloud, and with the PNTS tiles that are automatically exported out of DJI Terra, but if I generate tiles from the LAS with [gocesiumtiler](https://github.com/mfbonfigli/gocesiumtiler), it renders fine and in-position with the standard shader/style as simple RGB, but renders complete noise with the custom shader. 

The end goal is to change colour based on height in Cesium JS (I had no luck styling based on POSITION_ABSOLUTE so I looked to custom shaders)

Is there something wrong with my gocesiumtiler settings, or is this a bug? Can you offer any improvements to my gocesiumtiler parameters for a 3GB LAS?

[Sandcastle with custom shader](https://sandcastle.cesium.com/#c=lVZZb9s4EP4rRLAP8q5CS5Z85GixgesmQZI2iN30xYBDS7TNLUV6KcpHivz3HZKS4yMtsjZgkxzO/c1w6vV7yYRGXS6LFI3X6F7JCUb3VFOFLjinwkddyYtszAj6JtiCqpzpNXqQY6lZkqNbMsaonxAhmJgaARec/kNEqiQaKJkkkjMEW3RHNGWoy2RCFGcUD0UiRa7RgtElqPqABF2iLs1ZkeFHe+YNjxK770qhCRNUDY989HMoEALjFJycVgwDt8cTJbPvUvG0PPBq/lC81M6GYigWRKEZWML3tfUTRanoz0lCewsq9JW75DnLcJ5QQTE4uCC5kVTKwDnV12Je6ItEMym8SSHcghoZNWcnQkYreKxBFRGgd0fonCU/7iXEc8OH5+XWaDL8JkiSU8zl1NvIMcQX/1f2D9Zzim97nwej7u1198Z5D/7vupNRRYwPJtaeNTaluWaCGO2n2/HpVmojz9kUx0HYjmPciuMwbkZh7LvzRqMZNEMch0G7cxKGQbNREuLmSTNuBDhqx+1OHEeROa5ZolQMTD5UekVJCoC6ZzqZPUjOS9VNUNpqR1EnaAbRSbvZLjUcB7jRbrWCdtiMOtFJp9MqCS3c6IRgS6fdjIMg6DQr1S9VXMwXlR+TrYki0wxs6s9IStWArjSk7QniJ1m6Id4ZcH0uNxYGaJLbfx8xIWGbPGejTKaUG+ArRjjKykVtKCDeQ13tccomkyKnoCZjK28xSlk+52Rt2Ue5XnMIxAgKSSof7TP5KMChxcNQPJ3t+wLVqulq3xNXeBMuiUaXDxefrntfBqOr3vXl1QAuRAEOKsjQJEKzfNFQ07FnN4kF9lDDJkY3Bs+w8EIc+KiBA1RHkVmGr0v4sYHWln0OHGScexDGRHsJXq1W6C90g1fr5xr6E5IVoGPYLpdLx6SoLpRACX4GqgnOjWHxUcJJNvfm9rI9CEq1NaDhdRmPMmcuCDZjj3bp8rX4Rb7cna+Fdpfcoqxn64SiE6qoSKjtnHddF4XIszbYn2ZQel2xVFXtLjvFmGit2LiA2sKv9JLLJQeAoAkoKqO2rxiicrwlGj/XdrhnhWH0wCmvEuTv57tWg0TtnZ29ejodg4RtAJhUg1IQbcPtYu561e9wC1JAwllVbFVUwW9wpM+ejaENC7sDGDuwJkWuZeaAvNu9u1sU7+crH0KcTWfQ0KZ3Jqubp+J2+xR/+3J7PfC3uQrBJlJl+SnaEQaE0XKad+Kvgn4FRD1Ae2L9fwuiaHpwFR4o6MEbld+cSNuXH3vdyN+/vSC8eL3e45zNc0Au/n7Z78RYHmrck/Cys9/dQRtYg7uH/tTrqO9yhFyO9AxAw3J4rnhScOgzKZRGWT0od6E3LzmZz+EBt8SqHZbkXQW/A8TG2Udn3tux2fNkr5mdHpzsXD/s46dvnG2xlC+CVmsXKoc8zTiFZ9KU4JIwvYGd/Ys+DRzZTh7XUlzkZixIvbAVN0JXF6UAvAfh7e2ZScb58f/8lInIEUFm1hnLJZoqAAn4Z/TW65VmE3m6VzVb5hsUUM9eqr3bki3la8o56LblhpZMzwAYgFFb6DvTjmIZtKoF9DuSpl5pne3VgDl455EHU5tUtdfwV5PPU88QELQ1MxRUIT1Ff/y0HC9PTsq7xpx3DTqHI85b083hYFMr8fTuseZwoHlrljkcY2rlBHPkH53b1H2sgPw3y+ZSaVQo7mFc1zSbm2LO6+Mi+WFwmOfVcHle32Y9T9kCsfTDG2O3eXDzHCiTgnPTr4dHH8/rcP+AtUyRaVhQ+ebaLPx46w4xxud12L7NqaXkY6L2JP8H): ✅

![image](https://github.com/mfbonfigli/gocesiumtiler/assets/37650/df341c36-aee3-478b-89ed-793d45fb0a89)

DJI Terra generated tiles: ✅

![image](https://github.com/mfbonfigli/gocesiumtiler/assets/37650/c7691493-5861-4354-b023-941ea7c5b42a)

Same dataset exported as LAS from DJI Terra and tiled with gocesiumtiler: 👎

```sh
gocesiumtiler file \
  --epsg 32756 \
  --resolution 1 \
  --depth 8 \
  --min-points-per-tile 5000 \
  --out "tiles-gocesiumtiler" \
  "my-point-cloud.las"
```

![image](https://github.com/mfbonfigli/gocesiumtiler/assets/37650/146ef0d9-685e-43b0-a23e-65e5e2bad140)

Note that the positions of both tilesets are the same, I have used this method of generating tiles for quite a while but I’ve only ran into problems now that I am using a custom shader.

**Comment by mfbonfigli** [comment]
I am not quite sure what is causing it considering that from what you say the actual position of the points in space is correct. 

Would it be possible for you to share with me, even privately at my email address, the dataset you are using (or even just a small subset) for me to investigate further? 

It would also help if you could attach an example of a "working" tileset for reference.

**Comment by daz** [comment]
No worries, I've just email you a link with the dataset.

**Comment by mfbonfigli** [comment]
Thanks a lot. I believe that the issue here is that gocesiumtiler never uses local coordinate reference systems. All coordinates are stored in the Cesium default CRS EPSG 4978 (ECEF, cartesian geocentric reference system). Apparently the shader receives the raw coordinates prior to the global transforms, hence your shader will receive EPSG 4978 coords when reading the gocesiumtiler artifact, and coordinates in a local CRS when reading Terra models, hence the math is not going to be the same.

This means that you can't simply subtract the Z of a point to a baseline, as the Z axis is not "UP" relative to your model.

I managed to get a better visualization (not artifact free but almost) by rotating the points so that the Z is "up":

```
const float GRADIENT_HEIGHT = 30.0;


vec3 hsv2rgb(vec3 c) {
	vec4 K = vec4(1.0, 2.0 / 3.0, 1.0 / 3.0, 3.0);
	vec3 p = abs(fract(c.xxx + K.xyz) * 6.0 - K.www);
	return c.z * mix(K.xxx, clamp(p - K.xxx, 0.0, 1.0), c.y);
}

vec3 applyTransformation(mat4 transform, vec3 point) {
    // Convert the 3D point to a 4D point (homogeneous coordinates)
    vec4 point4D = vec4(point, 1.0);
    
    // Multiply the transformation matrix by the 4D point
    vec4 transformedPoint4D = transform * point4D;
    
    // Convert the 4D result back to a 3D point by dividing by w (in case w is not 1)
    return transformedPoint4D.xyz / transformedPoint4D.w;
}

void vertexMain(VertexInput vsInput, inout czm_modelVertexOutput vsOutput) {
 // vec3 positionMC = rotateToZAxis(vsInput.attributes.positionMC);

  mat4 rotationMatrix = mat4(
    -0.479541502619441,-0.8775192005110137,0.0,0.0,
    -0.35640222084710366,0.1947645777122581,0.9138075378567514,0.0,
    -0.8018836600409944,0.43820863980879826,-0.40614749015127727,0.0,
    -5117458.849367955,2796558.644859263,-2574599.756146693,1.0
  );
  vec3 positionMC = applyTransformation(inverse(rotationMatrix), vsInput.attributes.positionMC);
  
  vec3 referencePointMC = vec3(0.0, 0.0, 50 );

  float distance = abs(referencePointMC.z - positionMC.z);

  float hue = (mod(distance, GRADIENT_HEIGHT)) / GRADIENT_HEIGHT;
  vec3 rgb = hsv2rgb(vec3(1.0 - hue, 1.0, 1.0));
  v_display_mode_styling_color = rgb;


  vsOutput.pointSize = 2.0;
}
```

Note that I took the transform matrix from the terra tileset.json you sent me.

Unfortunately this is hard to fix on your side. You could tweak with the transform, but this also means that you need to store the transform matrix for each model somewhere or figure out a way to compute it dynamically.

I think this is a good use case. I will consider changing the tiler to store the points using a local "Z-up" reference with a global transform, this should help with these kind of tasks. I'm not sure on the effort required on my side to implement this change though but I believe it might be quite a lot.

**Comment by daz** [comment]
Thanks for putting the work in. I'd be happy to manually store the transformation for each tileset, providing I can get it perfect, but I was still getting artifacts with the rendering; It’s sort of placing triangle tiles rather than the squares that I’m used to getting out of ContextCapture, And the tiles are showing different colours in places they shouldn’t be. Is there something I could mitigate by tweaking the transformation matrix? I’ve attached pictures so the community can see the results. 

I would just use Terra tileset but I need to edit the LAS prior to tiling. Any other ideas? Changing the Coordinate system that Terra exports the LAS? Converting the Coordinate system of the LAS?

DJI Terra tiles:
![IMG_7030](https://github.com/mfbonfigli/gocesiumtiler/assets/37650/8c771032-f5f6-4e65-98d4-a3bc80a119eb)

Applying mfbonfigli’s transformation in the shader to the gocesiumtiler tiles, same data set but using the LAS from Terra:
![IMG_7029](https://github.com/mfbonfigli/gocesiumtiler/assets/37650/a7b7600a-1c04-4c0f-b12c-6579a736991d)

**Comment by mfbonfigli** [comment]
I am planning to provide a solution for this in the next release by shifting the coordinate storage to a local CRS. I already tested the preliminary code locally with the sample data you provided and is effective at preventing this issue.

**Comment by mfbonfigli** [comment]
Hey daz, I have released a new version 2.0.1 which should help you achieve what you need. Please check it out, I'll keep this issue open for a few days so you can provide feedback, then I'll eventually close it.

**Comment by mfbonfigli** [comment]
Closing this, feel free to reopen if the issue persists!
